### DISCUSSION TITLE
In Rust regex, is there a similar function to filtered_re.h in RE2

### DISCUSSION THREAD

**Discussion by yangwentong** [rust-lang/regex#923] [Q&A] [answered]
In RE2, [filtered_re.h](https://github.com/google/re2/blob/main/re2/filtered_re2.h) provides a prefilter mechanism that helps in cutting down the number of regexps that need to be actually searched.
Like:
``` C++
// filtered_re.h

  // Prepares the regexps added by Add for filtering.  Returns a set
  // of strings that the caller should check for in candidate texts.
  // The returned strings are lowercased and distinct. When doing
  // string matching, it should be performed in a case-insensitive
  // way or the search text should be lowercased first.  Call after
  // all Add calls are done.
  void Compile(std::vector<std::string>* strings_to_match);
```
I would like to ask if there is a function similar to `Compile` in Rust regex, or other functions provided in filtered_re.h?

**Comment by BurntSushi** [comment, marked as answer]
No. You can use the `regex-syntax` to look at the HIR and extract substrings yourself though if you want. And the `regex` crate does that internally and tries to do at least part of the `filtered_re.h` optimization on its own, automatically.

**Reply by yangwentong** [reply]
Thank you for answering my questions

**Comment by masklinn** [comment]
If anyone swings by and is still interested, I've given the port a shot: https://crates.io/crates/regex-filtered

It seems to work fine so far, though I won't discount having missed something (the underlying model of `re2` and `regex` are not quite the same, and `re2` makes use or some C++ things which are not available in rust so had to be worked around), and I've yet to profile it or bench it against `re2`.

Unlike `re2` the prefilter is built-in: since `regex` depends on `aho-corasick` which seems like the best prefilter you could ask for, it seems natural to just use it.

**Reply by BurntSushi** [reply]
This is awesome! Thanks for sharing!

Does `regex-filtered` have a repository? I don't see it linked in the crate metadata.

**Reply by masklinn** [reply]
Yeah I kinda fucked up and forgot to fill in those fields before I published it, I tried to re-publish but obviously cargo told me 0.1.0 was already pushed, I figure I'll have to bump to 0.1.1 in order for it to take.

The code is at https://github.com/ua-parser/uap-rust/tree/main/regex-filtered

**Reply by BurntSushi** [reply]
Yeah you'd need to republish. Crates are immutable once published. (The only exception to that, AFAIK, are legal issues.)
